### ISSUE TITLE
A question about DWEN

### ISSUE THREAD

**Issue by happygirlzt** [irving-muller/soft_alignment_model_bug_deduplication#6]
Dear @irving-muller, sorry to bother you again ( I hope this is the very last time :(
I have two questions related to DWEN
1. I can see that for all the other approaches, the training pairs, you use `training_BTS_pairs_random_1.txt` while for DWEN, you use `training_BTS_pairs_random_5.txt`. Is this a typo or just as it is? Why DWEN use a different number of training pairs
<img width="1423" alt="Screenshot 2021-07-06 at 16 01 20" src="https://user-images.githubusercontent.com/15061984/124566395-95298c00-de75-11eb-8331-b6ca0f8649ca.png">
2. For glove embeddings, it seems that only DWEN used `glove_300d` instead of `glove_42B_300d`, while for other approaches, `glove_42B_300d` is used. I wonder whether it is a typo or it should be like this. 
<img width="626" alt="Screenshot 2021-07-06 at 16 15 36" src="https://user-images.githubusercontent.com/15061984/124566905-0bc68980-de76-11eb-85e8-5a8b2ea8b024.png">
<img width="623" alt="Screenshot 2021-07-06 at 15 58 42" src="https://user-images.githubusercontent.com/15061984/124566916-0e28e380-de76-11eb-8a11-8b95d976e213.png">

Thank you very much for your patience and great help all the time. Have a nice week ahead! ❤️

**Comment by irving-muller** [comment]
Hello @happygirlzt , 

no worries. :)

1. It is correct. I found that DWEN achieved better results when the rate between positive and negative pairs were 5. `training_BTS_pairs_random_1.txt` and `training_BTS_pairs_random_5.txt` contains the same positive pairs and the only difference are the negative ones. However, this should no change the results since new negative pairs are sampled for each new epoch.
2. It is correct the file. glove_42B_300d is composed of glove_300d and [glove.42B.300d ](https://nlp.stanford.edu/data/glove.42B.300d.zip). Since  word embeddings were fine-tuned in DWEN, you only used the tokens (glove_300d) that appeared in the training.

**Comment by happygirlzt** [comment]
Dear @irving-muller, 
Thanks a lot for your reply :)
1. I see. Since it will not cause many changes, I think it would be fine I use the equal number of positive and negative pairs, i.e., the rate is 1, for DWEN too.
2. Oh, I did not realize this difference before. If you do not mind, I would like to ask for the details. For the GloVe embedding generation, I use the [GloVe official's repo](https://github.com/stanfordnlp/GloVe) and your guidance among the comments in `create_dataset_word_embedding.py` Like the following picture shows:
<img width="772" alt="Screenshot 2021-07-07 at 15 36 51" src="https://user-images.githubusercontent.com/15061984/124719044-50ffbf80-df39-11eb-8025-af6a0b62144b.png">


It seems that the generated files are named as
`glove_42b_300d_open_office_2001-2008_2010`.  If I am not wrong, the genrated files are in the red box in the following picture. 
<img width="1074" alt="Screenshot 2021-07-07 at 15 38 31" src="https://user-images.githubusercontent.com/15061984/124719255-899f9900-df39-11eb-8ea8-f9d5f44797a2.png">


Previously, I use these generated files as the 
 `glove_42B_300d_open_office_2001-2008_2010_soft_clean_merged`, I wonder whether it was wrong. If so, how to get the separate `glove_300d` and merged one? Many thanks for your time :) 
 
 I sincerely appreciate all your patient replies. Have a nice day ❤️

**Comment by irving-muller** [comment]
Hello @happygirlzt ,

1. When I said about "this should no change the results", I was mentioning about _the difference of the negatives pairs_ in the training files. I recall that the rate of 5 was significantly better than 1 for DWEN. 
2. The `SAVE_FILE` parameter value is not correct in the comment. The correct file name pattern is `glove_300d_dataset_soft_other_preprocessings.*`. `glove_42B_300d_open_office_2001-2008_2010_soft_clean_merged` is the combination of `globe_42B` and `glove_300d_dataset_soft.txt`

Sorry for the misunderstandings. Let me know if you have any question or there is something that is not clear.

**Comment by happygirlzt** [comment]
Dear @irving-muller , thanks a lot for your reply. 
> Sorry for the misunderstandings.

Don't worry, you are doing well. I think this is just because of my poor understanding:(

1. OK, got it, still need to keep the ratio of 5 then.
2. I see, I did make mistakes here. However, I'm not quite sure how to "combine" `glove.42B.300d.txt` (i.e., downloaded from GloVe website) and `glove_300_BTS_soft.txt`. Did you append one txt file to the other file? I'm wondering whether it makes sense. These two vector txt files seem to exist in different vector spaces. 
3. Also related to embeddings, like the figure below, why do `eclipse` and `netbeans` have the same suffix? While `open_office` and `mozilla` have the other set of the same suffix? Is there any reason here? They used two different cleaning strategies or these files are just samples here.
<img width="993" alt="Screenshot 2021-07-09 at 17 39 29" src="https://user-images.githubusercontent.com/15061984/125058188-9e686200-e0dc-11eb-97df-ec5bd30561df.png">

Many thanks again and have a nice Friday and weekend! ❤️

**Comment by irving-muller** [comment]
Hello @happygirlzt,

> I see, I did make mistakes here. However, I'm not quite sure how to "combine" glove.42B.300d.txt (i.e., downloaded from GloVe website) and glove_300_BTS_soft.txt. Did you append one txt file to the other file? I'm wondering whether it makes sense. These two vector txt files seem to exist in different vector spaces.

[This script](https://github.com/irving-muller/soft_alignment_model_bug_deduplication/blob/master/util/merge_wv_files.py) merges these files. Basically, I create a new file from `glove.42B.300d.txt` and I only add words from `glove_300_BTS_soft.txt` to it when they are not in `glove.42B.300d.txt`. In the beginning, I was also skeptical for the same reason but, empirically, this combination improved the performance because it reduces the number of unknown words. Since we trained the NN using these embeddings, I believe the NN can adapt to these two different dimension spaces.

>  Also related to embeddings, like the figure below, why do eclipse and netbeans have the same suffix? While open_office and mozilla have the other set of the same suffix? Is there any reason here? They used two different cleaning strategies or these files are just samples here.

I think it is a typo and  the same preprocessing steps were applied to textual data. However, I cannot confirm this since I lost the scripts used to train these word embeddings. So, I do not recommend to use them. 

It is important to highlight that I ran some experiments with different preprocessings for DWEN, but none of them significantly improved  its performance. In the end, I only used the files `glove_300d_BST_soft.txt` in the experiments.
